pkg/apis/core/validation/validation.go

```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
 	return allErrs
 }
 
+func validateWindowsHostProcessPod(podSpec *core.PodSpec, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	containerCount := 0
+	containersWithHostProcessSetTrue := []string{}
+	containersWithHostProcessSetFalse := []string{}
+	containersWhichInheritHostProcessTrue := []string{}
+
+	var podHostProcess *bool
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.WindowsOptions != nil && podSpec.SecurityContext.WindowsOptions.HostProcess != nil {
+		podHostProcess = podSpec.SecurityContext.WindowsOptions.HostProcess
+	}
+
+	hostNetwork := false
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.HostNetwork == true {
+		hostNetwork = true
+	}
+
+	// Note: Pod-wide WindowsSecurityContextOptions are inherited by containers that do not explicitly set them.
+	for _, c := range append(podSpec.Containers, podSpec.InitContainers...) {
+		containerCount++
+		if c.SecurityContext != nil && c.SecurityContext.WindowsOptions != nil && c.SecurityContext.WindowsOptions.HostProcess != nil {
+			if *c.SecurityContext.WindowsOptions.HostProcess {
+				containersWithHostProcessSetTrue = append(containersWithHostProcessSetTrue, c.Name)
+			} else if !*c.SecurityContext.WindowsOptions.HostProcess {
+				containersWithHostProcessSetFalse = append(containersWithHostProcessSetFalse, c.Name)
+			}
+		} else if podHostProcess != nil && *podHostProcess {
+			containersWhichInheritHostProcessTrue = append(containersWhichInheritHostProcessTrue, c.Name)
+		}
+	}
+
+	hostProcessContainerCount := len(containersWithHostProcessSetTrue) + len(containersWhichInheritHostProcessTrue)
+	if hostProcessContainerCount > 0 {
+		// Fail Pod validation if feature is not enabled instead of dropping fields based on PRR reivew.
```

Comment:
> // Fail Pod validation if feature is not enabled instead of dropping fields based on PRR review.

Do you have a pointer to where this was discussed? The current validation code fails validation of existing persisted pods using `hostProcess:true` if the feature gate is disabled, preventing updates needed to do things like remove finalizers.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version is the recommended approach to adding new alpha fields without breaking skewed API server validation.

Reply:
Part of the discussion happened here https://github.com/kubernetes/enhancements/pull/2288#discussion_r572301293 and the rest of the discussion happened on slack at https://kubernetes.slack.com/archives/CPNHUMN74/p1612805151060100.

Initially I was planning on dropping the fields like in the api_changes reference you linked but during the KEP review this was changed so as to not get into a situation where contianers that were meant to be scheduled as privileged/HostProcess containers would never get scheduled as regular containers if the fields got dropped.

Reply:
hmm... that differs from the typical approach... I'm trying to think if that causes any issues

to fix the "invalidates persisted data" issue, rather than checking feature gates inside validation, add an `AllowWindowsHostProcessField bool` field to `PodValidationOptions` and populate it inside `GetValidationOptionsFromPodSpecAndMeta`:

```diff
diff --git a/pkg/api/pod/util.go b/pkg/api/pod/util.go
index 59f1a198e75..43c30d69751 100644
--- a/pkg/api/pod/util.go
+++ b/pkg/api/pod/util.go
@@ -402,6 +402,8 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 		AllowInvalidPodDeletionCost: !utilfeature.DefaultFeatureGate.Enabled(features.PodDeletionCost),
 		// Do not allow pod spec to use non-integer multiple of huge page unit size default
 		AllowIndivisibleHugePagesValues: false,
+		// Allow hostProcess field to be set if the feature is enabled
+		AllowWindowsHostProcessField: utilfeature.DefaultFeatureGate.Enabled(features.WindowsHostProcessContainers),
 	}
 
 	if oldPodSpec != nil {
@@ -419,6 +421,8 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 
 		// if old spec used non-integer multiple of huge page unit size, we must allow it
 		opts.AllowIndivisibleHugePagesValues = usesIndivisibleHugePagesValues(oldPodSpec)
+		// if old spec has hostProcess containers, allow them in the update request
+		opts.AllowWindowsHostProcessField = opts.AllowWindowsHostProcessField || setsHostProcessField(oldPodSpec)
 	}
 	if oldPodMeta != nil && !opts.AllowInvalidPodDeletionCost {
 		// This is an update, so validate only if the existing object was valid.
diff --git a/pkg/apis/core/validation/validation.go b/pkg/apis/core/validation/validation.go
index 1b668083b3b..5d4a88d6100 100644
--- a/pkg/apis/core/validation/validation.go
+++ b/pkg/apis/core/validation/validation.go
@@ -3204,6 +3204,8 @@ type PodValidationOptions struct {
 	AllowInvalidPodDeletionCost bool
 	// Allow pod spec to use non-integer multiple of huge page unit size
 	AllowIndivisibleHugePagesValues bool
+	// Allow hostProcess field to be set in the windows security context
+	AllowWindowsHostProcessField bool
 }
 
 // ValidatePodSingleHugePageResources checks if there are multiple huge
```

`setsHostProcessField` should be implemented to return true if the pod-level or any container-level `hostProcess` field is non-null

plumb PodValidationOptions to `validateWindowsHostProcessPod`, and disallow any non-nil `hostProcess` field value (true or false) if `AllowWindowsHostProcessField` is false

Reply:
Thanks! I'll try this out (but probably not tonight...)

Reply:
Alright I think I got all of this coded up. Thanks for the guidance here. This is new to me.